### DISCUSSION TITLE
Create and add `Element` programmatically to the view

### DISCUSSION THREAD

**Discussion by Robbendebiene** [aurelia/aurelia#2034] [Q&A] [answered]
Hi everyone, my goal is to implement infinite scrolling in aurelia 2.
As far as I understand I cannot use the virtual repeat because the items have different heights. I'm not eager to use it anyway as I would just add a scroll listener and add items to the bottom of the list.

The problem for me is, how do I add an aurelia element with bindable properties to the view?
In the template file with aurelia I usually would do this:
```html
<post-view repeat.for="post of posts" post.bind="post"></post-view>
```
But how can I add the element programmatically?

Very basic working principle:

```ts
attached() {
  this.host.addEventListener('scroll', this.handleScroll, { passive: true });
}

private handleScroll() {
  if (this.isBottom) {
      this.host.append(post-view); // this is where I need help
  }
}
```

How do I create and add an `Element` (in this case `post-view`) to the DOM and bind properties to it?

Thanks in advance.

**Comment by bigopon** [comment]
@Robbendebiene if your template looks like this
```html
<post-view repeat.for="post of posts" post.bind="post"></post-view>
```
Maybe it's better to let Aurelia deal with view appending?
```html
<post-view repeat.for="post of visiblePosts" post.bind="post"></post-view>
```
```ts
attached() {
  this.host.addEventListener('scroll', () => {
    // logic to figure out new visible post here
    this.visiblePosts.push(...)
  }, { passive: true });
}
```

**Comment by Robbendebiene** [comment]
@bigopon Thanks for the quick reply.
This was actually my backup plan. Since I may have to discard elements at the top I tried to avoid using an array here.

To think from a different point of view, how hard is it to make a custom collection (like a LinkedList) work with `repeat.for=""`? Is there any interface that I can implement to make it work?

**Reply by bigopon** [reply, marked as answer]
You can specify your own collection handling strategy, doc is at https://docs.aurelia.io/templates/repeats-and-list-rendering#custom-collection-handling-strategy

Note that this only teaches the repeater how to iterate, it doesn't deal with observation so the repeater won't know when the linked list change

**Reply by Robbendebiene** [reply]
Thanks for the link. Any way to make it observable as well? 🫣

**Reply by bigopon** [reply]
By default, Aurelia cannot and should not know how to observe all kinds of mutation that can happen. It's only equipped to deal with most common mutations. You can teach the observer locator how to deal with it via the adapter API on it.
Though maybe this is a good question, we could try to make it easier to allow any collection to be easily hooked/observed. Either via the Symbol.iterator or some other way.

**Reply by Robbendebiene** [reply]
Sorry for the late reply.

Last weekend I finally could spent some time looking into the implementation of a custom observable collection. Unfortunately without any success so far.

At first I tried copying existing implementations like the [array observer](https://github.com/aurelia/aurelia/blob/c5dfca356050840cc36f68d212803db52bd71c84/packages/runtime/src/array-observer.ts) or [set observer](https://github.com/aurelia/aurelia/blob/c5dfca356050840cc36f68d212803db52bd71c84/packages/runtime/src/set-observer.ts).
Overriding the collection methods and mixing in a `o.notify();` felt familiar to me.

However I encountered two problems:
- not everything used in these implementations is exported/public. E.g.  `addCollectionBatch`, `batching`, `atObserver`, `rtDefineHiddenProp` 
- `ObservedCollectionKindToType` is mandatory, but all the possible values are pre-defined so this somehow seems to block the ability to create a custom collection

So I looked closer into the `IObjectObservationAdapter` as I thought this is what you meant by

> You can teach the observer locator how to deal with it via the adapter API on it.

I wasn't able to find any example implementation for it and since I have to return an `ICollectionObserver` I seem to run into the same problem as before.

I really would like to implement a custom observable collection as everything else I explored so far decreased the coded quality and complicated multiple things. (a strength of a linked list is to move items between lists but in this case both lists have to be notified of a change)

I would be very grateful if you could provide me a rough implementation of a basic observable collection via adapter.

______
[Mobx](https://mobx.js.org/) provides an `Atom` class which basically every observable must incorporate. When implementing a custom observable, one only hast to call the respective methods on the atom like `this.atom.reportChanged()` to make the reactive system work ([see example](https://mobx.js.org/custom-observables.html)). Aurelia is certainly taking a different root, so this is just an example of how other libraries make this possible.
